src/simulations/collect.jl

```diff
@@ -0,0 +1,285 @@
+  length(available_keys) == 0 && return
+  v1 = aggregation_dict[available_keys[1]]
+  final_df = by(df, :step,  available_keys[1] => v1[1])
+  for v2 in v1[2:end]
+    dd = by(df, :step,  available_keys[1] => v2)
+    final_df = join(final_df, dd, on=:step)
+  end
+  for k in available_keys[2:end]
+    v = aggregation_dict[k]
+    for v2 in v
+      dd = by(df, :step,  k => v2)
+      final_df = join(final_df, dd, on=:step)
+    end
+  end
+
+  # rename columns
+  colnames = Array{Symbol}(undef, size(final_df, 2)) 
+  colnames[1] = :step
+  counter = 2
+  for k in available_keys
+    v = aggregation_dict[k]
+    for vv in v
+      colnames[counter] = Symbol(join([vv,"(", string(k), ")"], ""))
+      counter += 1
+    end
+  end
+  rename!(final_df, colnames)
+
+  return final_df
+end
+
+## step functions
+###################
+
+"used in _step!"
+function _step_agent_collect!(df::DataFrame, model::ABM, agent_properties,  aggregation_dict, step::Int)
+  if isnothing(aggregation_dict)
+    dft = collect_agent_data(model, agent_properties, step)
+    df = vcat(df, dft)
+  else
+    dfall = collect_agent_data(model, agent_properties, step)
+    dfa = aggregate_data(dfall, aggregation_dict)
+    df = vcat(df, dfa)
+  end
+  return df
+end
+
+"used in _step!"
+function _step_model_collect!(df::DataFrame, model::ABM, model_properties,  aggregation_dict, step::Int)
+  if isnothing(aggregation_dict)
+    dft = collect_model_data(model, model_properties, step)
+    df = vcat(df, dft)
+  else
+    dfall = collect_model_data(model, model_properties, step)
+    dfa = aggregate_data(dfall, aggregation_dict)
+    df = vcat(df, dfa)
+  end
+  return df
+end
+
+function _step!(model, agent_step!, model_step!, n::F, when; model_properties=nothing, aggregation_dict=nothing, agent_properties=nothing) where F<:Function
+  df_agent = DataFrame()
+  df_model = DataFrame()
+  ss = 0
+  while !n(model)
+    step!(model, agent_step!, model_step!, 1)
+    if ss in when
+      if !isnothing(model_properties)
+        df_model = _step_model_collect!(df_model, model, model_properties,  aggregation_dict, ss)
+      end
+      if !isnothing(agent_properties)
+        df_agent = _step_agent_collect!(df_agent, model, agent_properties,  aggregation_dict, ss)
+      end
+    end
+    ss += 1
+  end
+  return df_agent, df_model
+end
+
+function _step!(model, agent_step!, model_step!, n::Int, when;
+  model_properties=nothing, aggregation_dict=nothing, agent_properties=nothing)
+  df_agent = DataFrame()
+  df_model = DataFrame()
+  for ss in 1:n
+    step!(model, agent_step!, model_step!, 1)
+    if ss in when
+      if !isnothing(model_properties)
+        df_model = _step_model_collect!(df_model, model, model_properties,  aggregation_dict, ss)
+      end
+      if !isnothing(agent_properties)
+        df_agent = _step_agent_collect!(df_agent, model, agent_properties,  aggregation_dict, ss)
+      end
+    end
+  end
+  return df_agent, df_model
+end
```

Comment:
same story can be done in the `n` condition, where we change 
```julia
until(ss, n::Int, model) = ss in 1:n
until(ss, n::AbstractVector{Int}, model) = ss in n
until(ss, n::Function, model) = !n(model)
```
and we change all `for ss in 1:n` to `while until(ss, n, model) ... ss += 1; end` 

This will lead to having only **one** function, in which every condition is satisfied. See also #167, and the issue that led to #167